`parse_tt`: a few more tweaks

nnethercote opened the pull request:
r? @petrochenkov

nnethercote commented:
I think the perf effects here should be negligible.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit c09baac998fa0fc3600a8f7f8650c023e9ac8e60 with merge d1dd78813e5e9620703b973bec06c3a6a03c506a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c09baac998fa0fc3600a8f7f8650c023e9ac8e60","merge_sha":"d1dd78813e5e9620703b973bec06c3a6a03c506a"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5880936688?check_suite_focus=true)
Build commit: d1dd78813e5e9620703b973bec06c3a6a03c506a (`d1dd78813e5e9620703b973bec06c3a6a03c506a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5880936688?check_suite_focus=true"},"merge_sha":"d1dd78813e5e9620703b973bec06c3a6a03c506a"} -->

rust-timer commented:
Queued d1dd78813e5e9620703b973bec06c3a6a03c506a with parent e745b4ddbd05026c75aae4506aef39fdfe1603c5, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e745b4ddbd05026c75aae4506aef39fdfe1603c5&end=d1dd78813e5e9620703b973bec06c3a6a03c506a).

fbstj commented on compiler/rustc_expand/src/mbe/macro_parser.rs:
```diff
@@ -219,7 +209,7 @@ struct MatcherPos {
     /// against the relevant metavar by the black box parser. An element will be a `MatchedSeq` if
     /// the corresponding metavar decl is within a sequence.
-    matches: Lrc<NamedMatchVec>,
+    matches: Lrc<Vec<NamedMatch>>,
```
your commit comment mentions that `Lrc` isn't necessary but it is still here?

rust-timer commented:
Finished benchmarking commit (d1dd78813e5e9620703b973bec06c3a6a03c506a): [comparison url](https://perf.rust-lang.org/compare.html?start=e745b4ddbd05026c75aae4506aef39fdfe1603c5&end=d1dd78813e5e9620703b973bec06c3a6a03c506a).

**Summary**: 
- Primary benchmarks: 😿 relevant regressions found
- Secondary benchmarks: 😿 relevant regressions found


| | Regressions 😿 <br />(primary) | Regressions 😿 <br />(secondary) | Improvements 🎉 <br />(primary) | Improvements 🎉 <br />(secondary) | All 😿 🎉 <br />(primary) |
|:---:|:---:|:---:|:---:|:---:|:---:|
| count[^1] | 2 | 6 | 0 | 1 | 2 |
| mean[^2] | 0.3% | 1.1% | N/A | -0.3% | 0.3% |
| max | 0.4% | 1.2% | N/A | -0.3% | 0.4% |

[^1]: *number of relevant changes*
[^2]: *the arithmetic mean of the percent change*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

nnethercote commented:
The slight perf losses will be from the third commit. I guess I'll drop that one.

petrochenkov commented on compiler/rustc_session/src/parse.rs:
```diff
@@ -140,6 +140,12 @@ pub struct ParseSess {
+    /// (ignorable) lint error is emitted. Ideally the lint error would be a hard error, emitted
+    /// during parsing, and this table wouldn't be needed. But this has proven difficult; see
+    /// #40107 for details.
     pub missing_fragment_specifiers: Lock<FxHashMap<Span, NodeId>>,
```
https://github.com/rust-lang/rust/pull/95808 removes this field.

petrochenkov commented on compiler/rustc_expand/src/mbe/macro_rules.rs:
```diff
@@ -441,7 +441,8 @@ pub fn compile_declarative_macro(
     let parser = Parser::new(&sess.parse_sess, body, true, rustc_parse::MACRO_ARGUMENTS);
-    let mut tt_parser = TtParser::new(def.ident);
+    let mut tt_parser =
+        TtParser::new(Ident::with_dummy_span(if macro_rules { kw::MacroRules } else { kw::Macro }));
```
This is reasonable while we are still matching the macro definition against `$( $lhs:tt => $rhs:tt );+` reusing `macro_rules` machinery.

However, we really should stop pretending that `macro_rules! foo { ... }` or `macro foo (...) { ... }` are macro invocations.
This involves all the comments saying that, introducing a custom parser for the `$( $lhs:tt => $rhs:tt );+` syntax (and its `macro` 2.0 equivalent), and storing the output from such parser (the lowered macro representation) to crate metadata, so that macro definitions are not reparsed as raw token streams by every dependent crate.

petrochenkov commented:
Not much remains from this PR if https://github.com/rust-lang/rust/pull/95794/commits/cc1e815283cc9235cc1bccc72669f15282c5707f and https://github.com/rust-lang/rust/pull/95794/commits/3595a0d923b20ee8577fd0f7411e46988e73f57d are dropped.

nnethercote commented:
I updated and removed the commit involving `missing_fragment_specifiers`.

The perf results might have been a weird [blip](https://rust-lang.zulipchat.com/#narrow/stream/247081-t-compiler.2Fperformance/topic/perf.20run.20blip). Let's try again.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit edd7f2cdab424ff8ed64a84f784a5c1273bcb138 with merge f5c9fe83718933285dee8c5664d4f4c5ae540007...
<!-- homu: {"type":"TryBuildStarted","head_sha":"edd7f2cdab424ff8ed64a84f784a5c1273bcb138","merge_sha":"f5c9fe83718933285dee8c5664d4f4c5ae540007"} -->

rust-timer commented:
Queued f5c9fe83718933285dee8c5664d4f4c5ae540007 with parent 1f7fb6413d6d6c0c929b223e478e44c3db991b03, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1f7fb6413d6d6c0c929b223e478e44c3db991b03&end=f5c9fe83718933285dee8c5664d4f4c5ae540007).

rust-timer commented:
Finished benchmarking commit (f5c9fe83718933285dee8c5664d4f4c5ae540007): [comparison url](https://perf.rust-lang.org/compare.html?start=1f7fb6413d6d6c0c929b223e478e44c3db991b03&end=f5c9fe83718933285dee8c5664d4f4c5ae540007).

**Summary**: 
- Primary benchmarks: 😿 relevant regressions found
- Secondary benchmarks: 🎉 relevant improvement found


| | Regressions 😿 <br />(primary) | Regressions 😿 <br />(secondary) | Improvements 🎉 <br />(primary) | Improvements 🎉 <br />(secondary) | All 😿 🎉 <br />(primary) |
|:---:|:---:|:---:|:---:|:---:|:---:|
| count[^1] | 4 | 0 | 0 | 1 | 4 |
| mean[^2] | 0.4% | N/A | N/A | -0.2% | 0.4% |
| max | 0.4% | N/A | N/A | -0.2% | 0.4% |

[^1]: *number of relevant changes*
[^2]: *the arithmetic mean of the percent change*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

nnethercote commented:
Perf changes are negligible. I think the simplification of removing the `Lrc` use is worthwhile.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit edd7f2cdab424ff8ed64a84f784a5c1273bcb138 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov edd7f2cdab424ff8ed64a84f784a5c1273bcb138 -->
<!-- homu: {"type":"Approved","sha":"edd7f2cdab424ff8ed64a84f784a5c1273bcb138","approver":"petrochenkov"} -->

nnethercote commented:
Oh, I forgot to mark this as triaged: see [this comment](https://github.com/rust-lang/rust/pull/95794#issuecomment-1094515008)

@rustbot label: +perf-regression-triaged
